Jenny documentation

jennyfolkesson opened the pull request:
Hi Ziwen, 
I just added sphinx to the microDL directory so we can automatically document stuff. It looks like a big PR but it's really just all the sphinx files. I've only changed the source/conf.py file and some docstrings. If you could take a look I'd appreciate it!

ziw-liu commented:
Some cosmetic comments:

<img width="1101" alt="Screen Shot 2022-12-07 at 12 01 06" src="https://user-images.githubusercontent.com/67518483/206283641-07a7171a-53d3-42c3-bad4-aa413c885de0.png">

I'm not familiar with the Sphinx template, but I guess the top-let 'microDL_docs' title is due to this line:
https://github.com/mehta-lab/microDL/blob/f539949b765e2a0bfd06e83903e29516c58d499f/source/conf.py#L14
and might look better if we just call it 'microDL'.

The duplication in 'Documentation's documentation' may be caused by the auto-generated line:
https://github.com/mehta-lab/microDL/blob/f539949b765e2a0bfd06e83903e29516c58d499f/source/index.rst#L6
but we should be fine to edit that since the auto-generation seems to be quick-start only according to the comments in the file.

Similarly it would be nice if we don't have the submodules be called 'Subpackages'.

ziw-liu commented:
It was not obvious to me how are we going to deploy this page. Will it be on [GitHub Pages](https://pages.github.com/), [Read the Docs](https://readthedocs.org/), or self-hosted? Are we considering a CD workflow for the docs?

jennyfolkesson commented:
Thank you @ziw-liu for your feedback! I've fixed the documentation's documentation, and microDL_docs issues now.
My plan was to host it on GitHub pages. The docstrings contain a lot of information and we might as well leverage that with sphinx and github pages since it's just a few commands that need to be called each time we make a new release.

mattersoflight commented:
Thanks for this PR @jennyfolkesson. @ziw-liu let's start with hosting via GitHub pages and we should consider read the docs. 

We will occasionally need to add tutorials and descriptions, e.g., @Soorya19Pradeep documented configs with micoDL 1.0.0. Does sphinx require specific configuration to mix auto-generated and hand-written docs?

ziw-liu commented:
> it's just a few commands that need to be called each time we make a new release

Great to know! Can you also document in this PR the sphinx build commands and arguments so that this workflow is reproducible? Also by 'CD workflow' I meant something like [this](https://github.com/marketplace/actions/sphinx-build#advanced-usage), which basically runs those build commands for us.

jennyfolkesson commented:
> > it's just a few commands that need to be called each time we make a new release
> 
> Great to know! Can you also document in this PR the sphinx build commands and arguments so that this workflow is reproducible? Also by 'CD workflow' I meant something like [this](https://github.com/marketplace/actions/sphinx-build#advanced-usage), which basically runs those build commands for us.

I can't find a way to automatically convert "subpackages" to "submodules", so I changed them in the rst's.

Now I see what you're saying. Great idea to automate the sphinx build and push to github actions (actions_docs.yml. I've added a workflow for it now, I don't know if it will work before this branch gets merged to main.
Please let me if there are any other changes I should make before you can approve the PR. Thanks again.

jennyfolkesson commented:
> Thanks for this PR @jennyfolkesson. @ziw-liu let's start with hosting via GitHub pages and we should consider read the docs.
> 
> We will occasionally need to add tutorials and descriptions, e.g., @Soorya19Pradeep documented configs with micoDL 1.0.0. Does sphinx require a specific configuration to mix auto-generated and hand-written docs?

By hand-written docs, do you mean any documentation occurring outside of docstrings?
I don't know of a direct way to add commented files like the config files to the sphinx documentation. Perhaps add a reference to the commented file in the docstring of the CLI that uses that file?

ziw-liu commented:
> a direct way to add commented files like the config files to the sphinx documentation

For this I think one way is to add RST files (one can write in Markdown and convert to RST) with code blocks.

mattersoflight commented:
Hi @jennyfolkesson and @ziw-liu Let's host the docs for release 1.0.0 (https://github.com/mehta-lab/microDL/releases/tag/1.0.0) on a `gh-pages` branch as recommended [here](https://www.docslikecode.com/articles/github-pages-python-sphinx/). I scanned the hand-written docs at that commit, and all of them are linked correctly in README. 
Please also document how to build the docs and save on `gh-pages` with every release.

jennyfolkesson commented:
@mattersoflight @ziw-liu I've added a brief readme to the docs directory now on how to build the docs. Please review the PR again. Thanks.

ziw-liu commented on .github/workflows/actions_docs.yml:
```diff
@@ -0,0 +1,39 @@
+        git commit -m "Update documentation" -a || true
+        # The above command will fail if no changes were present, so we ignore
+        # that.
+    - name: Push changes
```
Have you tested if this workflow will work?

jennyfolkesson replied:
@ziw-liu It was something I didn't test because I didn't see a need to figure out how to checkout this branch after cloning the repo to github pages, since in the future this action will only happen when pushing to main and we only want to clone the main branch.
So I removed the actions_docs from this PR. I will submit a separate PR with this workflow once this PR is merged to main. Thank you for reviewing again.
